[DataPipe] Count number of successful yields for IterDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* #79658
* #79479
* __->__ #79657

This PR adds an attribute and logic to count the number of successful yields from `IterDataPipe`. This information can be useful to fast-forward a DataPipe (or the entire graph) back to a certain state.

NivekT committed:
[DataPipe] Count number of yields for IterDataPipe

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of yields for IterDataPipe"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of yields for IterDataPipe"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of yields for IterDataPipe"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of yields for IterDataPipe"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of successful yields for IterDataPipe"




[ghstack-poisoned]

ejguan reviewed (commented):
I have two comments below about bumping and reducing the counter.

NivekT commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -94,26 +94,46 @@ def profiler_record_fn_context():
+            _check_iterator_valid(self.source_dp, self.iterator_id)
+            try:
+                # Avoid double counting since if `__next__` is defined, it will be called
+                if not self.self_and_has_next_method:
+                    self.source_dp._number_of_samples_yielded += 1
+                return next(self.iterator)
+            except Exception:
+                # If there is an exception,the element was not successfully returned, the count needs to -= 1.
+                if not self.self_and_has_next_method:
+                    self.source_dp._number_of_samples_yielded -= 1
+                raise
```
I have a different idea on how to prevent it. Let me give it a try and get back to you.

NivekT replied:
It is the same as your idea. Thanks1

NivekT committed:
Update on "[DataPipe] Count number of successful yields for IterDataPipe"


This PR adds an attribute and logic to count the number of successful yields from `IterDataPipe`. This information can be useful to fast-forward a DataPipe (or the entire graph) back to a certain state.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of successful yields for IterDataPipe"


This PR adds an attribute and logic to count the number of successful yields from `IterDataPipe`. This information can be useful to fast-forward a DataPipe (or the entire graph) back to a certain state.

[ghstack-poisoned]

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2588,5 +2588,120 @@ def test_iterdatapipe_singleton_constraint_multiple_outputs(self):
+            if i == n_expected_samples - 1:
+                break
+        self.assertEqual(n_expected_samples, datapipe._number_of_samples_yielded)
+
```
You can include reset operation test here by simply changing order of full traversal and partial traversal.

NivekT committed:
Update on "[DataPipe] Count number of successful yields for IterDataPipe"


This PR adds an attribute and logic to count the number of successful yields from `IterDataPipe`. This information can be useful to fast-forward a DataPipe (or the entire graph) back to a certain state.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Count number of successful yields for IterDataPipe"


This PR adds an attribute and logic to count the number of successful yields from `IterDataPipe`. This information can be useful to fast-forward a DataPipe (or the entire graph) back to a certain state.

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge
